File: src/com/jupiter/ganymede/math/regression/LinearRegressor.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.jupiter.ganymede.math.regression;

import java.util.Set;
import java.util.function.Function;

/**
 *
 * @author Nathan Templon
 */
public class LinearRegressor implements Regressor<Double, Double> {

    @Override
    public Function bestFit(Set<Point<? extends Double, ? extends Double>> points) {
        final double firstMean = points.stream()
                .mapToDouble((Point<? extends Double, ? extends Double> point) -> point.first)
                .average()
                .getAsDouble();

        final double secondMean = points.stream()
                .mapToDouble((Point<? extends Double, ? extends Double> point) -> point.second)
                .average()
                .getAsDouble();

        final double firstVariance = points.stream()
                .mapToDouble((Point<? extends Double, ? extends Double> point) -> (point.first - firstMean) * (point.first - firstMean))
                .sum() / (points.size() - 1);

        final double secondVariance = points.stream()
                .mapToDouble((Point<? extends Double, ? extends Double> point) -> (point.second - secondMean) * (point.second - secondMean))
                .sum() / (points.size() - 1);

        final double firstStdDev = Math.sqrt(firstVariance);
        final double secondStdDev = Math.sqrt(secondVariance);

        final double correlation = points.stream()
                .mapToDouble((Point<? extends Double, ? extends Double> point) -> point.first * point.second)
                .sum()
                / Math.sqrt(
                        points.stream()
                        .mapToDouble((Point<? extends Double, ? extends Double> point) -> point.first * point.first)
                        .sum()
                        * points.stream()
                        .mapToDouble((Point<? extends Double, ? extends Double> point) -> point.second * point.second)
                        .sum()
                );
        
        final double slope = correlation * (secondStdDev / firstStdDev);
        final double yIntercept = secondMean - slope * firstMean;
        
        return (Function<Double, Double>) (Double value) -> yIntercept + value * slope;
    }

}

File: src/com/jupiter/ganymede/math/regression/LinearRegressor.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.jupiter.ganymede.math.regression

import com.jupiter.ganymede.math.function.LinearFunction
import java.util.function.Function

/**
 * @author Nathan Templon
 */
public class LinearRegressor : Regressor<Double, Double> {

    public override fun bestFit(points: Set <Regressor.Point<Double, Double>>): LinearFunction {
        val firstMean = points.map { it.first }.average()
        val secondMean = points.map { it.second }.average()

        val firstVariance = points.map { (it.first - firstMean) * (it.first - firstMean) }.sum() / (points.size() - 1)
        val secondVariance = points.map { (it.second - secondMean) * (it.second - secondMean) }.sum() / (points.size() - 1)

        val firstStdDev = Math.sqrt(firstVariance);
        val secondStdDev = Math.sqrt(secondVariance);

        val correlation = points.map { it.first * it.second }.sum() / Math.sqrt(
                points.map{it.first * it.first}.sum() * points.map{it.second * it.second}.sum()
        )

        val slope = correlation * (secondStdDev / firstStdDev);
        val yIntercept = secondMean - slope * firstMean;

        if (slope.isNaN() || yIntercept.isNaN()) {
            System.out.println("NaN in Regressor");
        }

        return LinearFunction(slope, yIntercept);
    }
}
